File: src/main/java/de/pflugradts/passbird/domain/service/password/provider/RandomPasswordProvider.java
package de.pflugradts.passbird.domain.service.password.provider;

import de.pflugradts.passbird.domain.model.password.PasswordRequirements;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.model.transfer.CharValue;

import java.security.SecureRandom;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static de.pflugradts.passbird.domain.model.transfer.CharValueKt.FIRST_DIGIT_INDEX;
import static de.pflugradts.passbird.domain.model.transfer.CharValueKt.LAST_LOWERCASE_INDEX;
import static de.pflugradts.passbird.domain.model.transfer.CharValueKt.MAX_ASCII_VALUE;
import static de.pflugradts.passbird.domain.model.transfer.CharValueKt.MIN_ASCII_VALUE;

public class RandomPasswordProvider implements PasswordProvider {

    private final SecureRandom random = new SecureRandom();

    @Override
    public Bytes createNewPassword(final PasswordRequirements passwordRequirements) {
        Bytes passwordBytes = Bytes.emptyBytes();
        while (!isStrong(passwordBytes, passwordRequirements)) {
            passwordBytes = randomPassword(passwordRequirements);
        }
        return passwordBytes;
    }

    private Bytes randomPassword(final PasswordRequirements passwordRequirements) {
        return Bytes.bytesOf(IntStream.range(0, passwordRequirements.getPasswordLength())
                .mapToObj(i -> nextByte(passwordRequirements)).collect(Collectors.toList()));
    }

    private Byte nextByte(final PasswordRequirements passwordRequirements) {
        return passwordRequirements.getIncludeSpecialCharacters()
                ? randomByte(MIN_ASCII_VALUE, MAX_ASCII_VALUE + 1)
                : randomByte(FIRST_DIGIT_INDEX, LAST_LOWERCASE_INDEX + 1);
    }

    private byte randomByte(final int fromInclusive, final int toExclusive) {
        return (byte) (random.nextInt(toExclusive - fromInclusive) + fromInclusive);
    }

    private boolean isStrong(final Bytes passwordBytes, final PasswordRequirements requirements) {
        return anyMatch(passwordBytes.copy(), c -> CharValue.Companion.charValueOf(c).isDigit())
                && anyMatch(passwordBytes.copy(), c -> CharValue.Companion.charValueOf(c).isUppercaseCharacter())
                && anyMatch(passwordBytes.copy(), c -> CharValue.Companion.charValueOf(c).isLowercaseCharacter())
                && anyMatch(passwordBytes.copy(),
                    c -> CharValue.Companion.charValueOf(c).isSymbol()) == requirements.getIncludeSpecialCharacters();
    }

    private boolean anyMatch(final Bytes bytes, final Predicate<Byte> predicate) {
        final var result = bytes.stream().anyMatch(predicate);
        bytes.scramble();
        return result;
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/service/password/provider/RandomPasswordProvider.kt
package de.pflugradts.passbird.domain.service.password.provider

import de.pflugradts.passbird.domain.model.egg.PasswordRequirements
import de.pflugradts.passbird.domain.model.shell.MAX_ASCII_VALUE
import de.pflugradts.passbird.domain.model.shell.MIN_ASCII_VALUE
import de.pflugradts.passbird.domain.model.shell.PlainValue
import de.pflugradts.passbird.domain.model.shell.PlainValue.Companion.plainValueOf
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.emptyShell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import java.security.SecureRandom
import kotlin.reflect.KProperty1

private val random = SecureRandom()

class RandomPasswordProvider : PasswordProvider {
    override fun createNewPassword(passwordRequirements: PasswordRequirements): Shell {
        var passwordShell = emptyShell()
        while (!isStrong(passwordShell, passwordRequirements)) { passwordShell = randomPassword(passwordRequirements) }
        return passwordShell
    }

    private fun randomPassword(passwordRequirements: PasswordRequirements) =
        shellOf((0..<passwordRequirements.length).map { randomByte(passwordRequirements) }.toList())

    private fun randomByte(passwordRequirements: PasswordRequirements): Byte {
        val getRandom = { (random.nextInt(MAX_ASCII_VALUE - MIN_ASCII_VALUE) + MIN_ASCII_VALUE).toByte() }
        var result: Byte
        do { result = getRandom() } while (!result.satisfies(passwordRequirements))
        return result
    }

    private fun isStrong(passwordShell: Shell, passwordRequirements: PasswordRequirements) =
        (!passwordRequirements.hasNumbers || passwordShell.anyMatch(PlainValue::isDigit)) &&
            (!passwordRequirements.hasUppercaseLetters || passwordShell.anyMatch(PlainValue::isUppercaseCharacter)) &&
            (!passwordRequirements.hasLowercaseLetters || passwordShell.anyMatch(PlainValue::isLowercaseCharacter)) &&
            (!passwordRequirements.hasSpecialCharacters || passwordShell.anyMatch(PlainValue::isSymbol))

    private fun Shell.anyMatch(property: KProperty1<PlainValue, Boolean>) = copy().stream().anyMatch { property.get(plainValueOf(it)) }
    private fun Byte.matches(property: KProperty1<PlainValue, Boolean>) = property.get(plainValueOf(this))
    private fun Byte.satisfies(passwordRequirements: PasswordRequirements): Boolean {
        if (passwordRequirements.hasSpecialCharacters && this.toInt().toChar() in passwordRequirements.unusedSpecialCharacters) return false
        if (!passwordRequirements.hasSpecialCharacters && matches(PlainValue::isSymbol)) return false
        if (!passwordRequirements.hasNumbers && passwordRequirements.isValid() && matches(PlainValue::isDigit)) return false
        if (!passwordRequirements.hasLowercaseLetters && matches(PlainValue::isLowercaseCharacter)) return false
        if (!passwordRequirements.hasUppercaseLetters && matches(PlainValue::isUppercaseCharacter)) return false
        return true
    }
}
